mirrord-layer/src/pod_api.rs

```diff
@@ -412,8 +381,252 @@ async fn create_job_pod_agent(
+        Ok((pod_api, deployment_api))
+    }
+
+    fn merge(
+        &self,
+        response_data: Response,
+        container_data: ContainerRuntime,
+    ) -> Result<RuntimeData> {
+        let Response {
+            node_name,
+            pod_name,
+            container_name,
+            ..
+        } = response_data;
+
+        // TODO: any way to clear this mess?
+        let node_name = node_name.ok_or_else(|| LayerError::NodeNotFound(self.clone()))?;
+        let pod_name = pod_name.ok_or_else(|| LayerError::PodNotFound(self.clone()))?;
+        let container_name =
+            container_name.ok_or_else(|| LayerError::TargetContainerNotFound(self.clone()))?;
+
+        let ContainerRuntime {
+            socket_path,
+            container_id,
+            container_runtime,
+        } = container_data;
+
+        Ok(RuntimeData {
+            node_name,
+            pod_name,
+            container_name,
+            container_id,
+            socket_path,
+            container_runtime,
+        })
+    }
+}
+
+impl FromStr for DeploymentData {
+    type Err = LayerError;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let target_data = input.split('/').collect::<Vec<&str>>();
+        match target_data.first() {
+            Some(&"deployment") if target_data.len() == 2 => Ok(DeploymentData {
+                deployment: target_data[1].to_string(),
+            }),
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is not a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub(crate) struct PodData {
+    pub pod_name: String,
+    pub container_name: Option<String>,
+}
+
+impl FromStr for PodData {
+    type Err = LayerError;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let target_data = input.split('/').collect::<Vec<&str>>();
+        match target_data.first() {
+            Some(&"pod") if target_data.len() == 2 => Ok(PodData {
+                pod_name: target_data[1].to_string(),
+                container_name: None,
+            }),
+            Some(&"pod") if target_data.len() == 4 && target_data[2] == "container" => {
+                Ok(PodData {
+                    pod_name: target_data[1].to_string(),
+                    container_name: Some(target_data[3].to_string()),
+                })
+            }
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is neither a pod or a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+impl PodData {
+    pub async fn container_data(&self, pods_api: &Api<Pod>) -> Option<Response> {
+        let pod = pods_api.get(&self.pod_name).await.ok()?;
+        let container_statuses = &pod.clone().status?.container_statuses?;
+        // TODO: should the return type be a Result here?
+        let container_info = if let Some(container_name) = &self.container_name {
```

Comment:
So is `container_info` just the container ID?

Reply:
its runtime + container id